driver/parallel.go

```diff
@@ -0,0 +1,222 @@
+package driver
+
+import (
+	"sync"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/filter"
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+)
+
+type parallelHead struct {
+	proc.Base
+	once sync.Once
+	pg   *parallelGroup
+
+	mu sync.Mutex // protects below
+	sc ScannerCloser
+}
+
+func (ph *parallelHead) closeOnDone() {
+	select {
+	case <-ph.Context.Done():
+		ph.mu.Lock()
+		defer ph.mu.Unlock()
+		if ph.sc != nil {
+			ph.sc.Close()
+			ph.sc = nil
+		}
+	}
+}
+
+func (ph *parallelHead) Pull() (zbuf.Batch, error) {
+	// Trigger the parallel group to read from the multisource.
+	ph.pg.once.Do(func() {
+		go ph.pg.run()
+	})
+	// Ensure open scanners are closed when flowgraph execution stops.
+	ph.once.Do(func() {
+		go ph.closeOnDone()
+	})
+
+	ph.mu.Lock()
+	defer ph.mu.Unlock()
+
+	for {
+		if ph.sc == nil {
+			sc, err := ph.pg.nextSource()
+			if sc == nil || err != nil {
+				return nil, err
+			}
+			ph.sc = sc
+		}
+		batch, err := ph.sc.Pull()
+		if err != nil {
+			return nil, err
+		}
+		if batch == nil {
+			if err := ph.sc.Close(); err != nil {
+				return nil, err
+			}
+			ph.pg.doneSource(ph.sc)
+			ph.sc = nil
+			continue
+		}
+		return batch, err
+	}
+}
+
+type parallelGroup struct {
+	filter     SourceFilter
+	msrc       MultiSource
+	once       sync.Once
+	pctx       *proc.Context
+	sourceChan chan SourceOpener
+	sourceErr  error
+
+	mu       sync.Mutex // protects below
+	stats    scanner.ScannerStats
+	scanners []scanner.Scanner
```

Comment:
Nit: You could simplify the methods a little if you changed the type of scanners to map[scanner.Scanner]struct{}.

Reply:
oh, that's a good suggestion!